### PR TITLE
Add migration test for divulgence

### LINKED ISSUE #6349: Add migration test to verify that contract divulgence works across SDK updates
A test should be added to `//compatibility/sandbox-migration` to verify that in case of migration the visibility of divulged contracts is preserved across SDK updates.

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
Closes #6349

changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compatibility/sandbox-migration/src/com/daml/Divulgence.scala [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  LedgerIdRequirement
+}
+import spray.json.RootJsonFormat
+
+import scala.concurrent.{ExecutionContext, Future}
+
+object Divulgence {
+
+  private final case class Result(
+      oldPrivateAssets: Seq[Application.ContractResult],
```
Comment (cocreature) [inline compatibility/sandbox-migration/src/com/daml/Divulgence.scala:25]: The terminology here sounds a bit like a contract is either private or divulged but that doesn’t seem to be the case afaik. `oldPrivateAssets` are just all assets returned by the ACS query and the divulged ones are a subset of that. If that is the case and I’m not missing something, I would suggest to just drop the `private` prefix here and in the Haskell code.

File: compatibility/sandbox-migration/runner/Migration/Util.hs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+-- SPDX-License-Identifier: Apache-2.0
+
+module Migration.Util (symDiff, equivalent) where
+
+import Data.List
+
+symDiff :: Eq a => [a] -> [a] -> [a]
+symDiff left right = (left \\ right) ++ (right \\ left)
+
+equivalent :: Eq a => [a] -> [a] -> Bool
```
Comment (cocreature) [inline compatibility/sandbox-migration/runner/Migration/Util.hs:11]: fwiw if you have an `Ord` instance which you can just derive you should be able to turn this into `equivalent xs ys = sort xs == sort ys` which has a more reasonable complexity.

Comment (stefanobaghino-da) [reply compatibility/sandbox-migration/runner/Migration/Util.hs:11]: I know, I tried initially but adding `Ord` to the `deriving` clause on the relevant data types returned me an error that defied my knowledge of Haskell.

I just tried again and I probably just missed `deriving Ord` for `Party` as well. Thanks for raising this.

Comment (cocreature) [reply compatibility/sandbox-migration/runner/Migration/Util.hs:11]: As always, if you need help figuring out the error, I’m happy to help. Also happy to not change this as part of this PR.

File: compatibility/sandbox-migration/daml/Divulgence.daml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+module Divulgence where
+
+template Asset
+  with
+    owner: Party
+    tag: Text
+  where
+    signatory owner
+
+    nonconsuming choice Fetch: Asset
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply compatibility/sandbox-migration/daml/Divulgence.daml:13]: Wild guess: preventing contract information leakage do divulgees unless it was explicitly allowed by the model (as in this case)? @gerolf-da do you have something to share here?

File: compatibility/sandbox-migration/runner/Migration/Divulgence.hs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            , "--test=divulgence," <> T.unpack (getParty testOwner) <> "," <> T.unpack (getParty testDivulgee) <> "," <> suffix
+            ]
+        either fail pure =<< A.eitherDecodeFileStrict' outputFile
+    validateStep sdkVersion (prevPrivateAssets, prevDivulgedAssets) Result {..} = do
+        let suffix = getSdkVersion sdkVersion
+        unless (equivalent oldPrivateAssets prevPrivateAssets) $
+            throwError ("The old private assets do not match those returned by the previous run: " <> show oldPrivateAssets)
+        unless (equivalent oldDivulgedAssets prevDivulgedAssets) $
+            throwError ("The old divulged assets do not match those returned by the previous run: " <> show oldDivulgedAssets)
+        let privateAssetDiff = (map _2 newPrivateAssets) \\ (map _2 oldPrivateAssets)
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply compatibility/sandbox-migration/runner/Migration/Divulgence.hs:41]: I though `hlint` was run as part of `./fmt.sh --test`. How do I make sure to catch these before I push?

[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply compatibility/sandbox-migration/runner/Migration/Divulgence.hs]: For some reason I can't batch these two suggestions in GitHub, so I just made a new commit for those. Thanks for the suggestion.

File: compatibility/sandbox-migration/runner/Migration/Util.hs (hunk 2 above)
Comment (stefanobaghino-da) [reply compatibility/sandbox-migration/runner/Migration/Util.hs]: Address https://github.com/digital-asset/daml/pull/6376/commits/0de90238f5da1e55f38e42ea7b3aa0ea2662545f

Comment (stefanobaghino-da) [reply compatibility/sandbox-migration/runner/Migration/Util.hs]: > As always, if you need help figuring out the error, I’m happy to help. Also happy to not change this as part of this PR.

Thanks, I just didn't realize that I didn't have all the necessary type classes instances in place, not it's solved. :bow:

File: compatibility/sandbox-migration/src/com/daml/Divulgence.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply compatibility/sandbox-migration/src/com/daml/Divulgence.scala]: Makes sense, thanks. Addressed by https://github.com/digital-asset/daml/pull/6376/commits/fa910bde5a6d264a8e780055155e55e63136d442.

File: compatibility/sandbox-migration/runner/Migration/Divulgence.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply compatibility/sandbox-migration/runner/Migration/Divulgence.hs]: Addressed by https://github.com/digital-asset/daml/pull/6376/commits/c67b4e5b47eac37c4b401577706141f2217862e2

File: compatibility/sandbox-migration/runner/Migration/Divulgence.hs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        unless (equivalent privateAssetDiff [Asset testOwner ("private-" <> suffix), Asset testOwner ("divulging-" <> suffix)]) $
+            throwError ("Expected one private and one divulged contract, got " <> show privateAssetDiff)
+        let divulgedAssetDiff = (map _2 newDivulgedAssets) \\ (map _2 oldDivulgedAssets)
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply compatibility/sandbox-migration/runner/Migration/Divulgence.hs]: Addressed by https://github.com/digital-asset/daml/pull/6376/commits/c67b4e5b47eac37c4b401577706141f2217862e2

File: compatibility/sandbox-migration/runner/Migration/Divulgence.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cocreature) [reply compatibility/sandbox-migration/runner/Migration/Divulgence.hs]: I thought so too and I’m almost certain I did run that on files in the migration tests in the last couple of days and it did, did you maybe run it before committing? I think it only runs on files that are committed.

**Pull request merged**